### DISCUSSION TITLE
flutter-web-app audio autoplay restriction policy: play a sound at defined interval

### DISCUSSION THREAD

**Discussion by toutPetitPoissonRouge** [community/community#122934] [Programming Help]
### Body

Hello,

i want to create a web-app which will play a sound (from asset) every 2 s. You find below a simplifed version of the code. This code works perfectely fine on desktop browsers like chrome or firefox. But the audio fonction does not even start on mobile browser (chrome, safari). I guess it is due to the autoplay restriction policy in mobile browsers. Does any body faced the same issue ? and/or has ideas how to solve that ?  

Here is the code:

```

import 'package:flutter/material.dart';
import 'package:audioplayers/audioplayers.dart';
import `'dart:async';

void main() {
  runApp(const MyApp());
}
class MyApp extends StatelessWidget {
  const MyApp({super.key});
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: const Text('Test sound every 2 s'),
        ),
        body: const Buttonsound(),
      ),
    );
  }
}

class Buttonsound extends StatelessWidget {
  const Buttonsound({super.key});
  @override
  Widget build(BuildContext context) {    
    AudioPlayer player = AudioPlayer();
    late Timer timer;    
    return Row(
      mainAxisAlignment: MainAxisAlignment.center,        
      children: [                
        SizedBox(                    
         child: ElevatedButton(
            onPressed: () {                     
              timer = Timer.periodic(const Duration(milliseconds: 2000), (timer) async  {      
                player.play(AssetSource('audio/68221786.mp3'));       
              });
            },
            child: const Text('Play sound every 2 s'),
          ),
        ),
      ],
    );
  }
}
```

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by chienhsiang-hung** [comment]
You're correct that mobile browsers have restrictions on autoplaying audio to prevent unsolicited content from playing without a user's interaction. This is likely the reason why your audio function isn't starting on mobile browsers.

To work around the autoplay restrictions on mobile browsers, you can consider the following approaches:

- **User Interaction:** Initiate the audio play through a user action, such as a button press or a touch event.
- **Silent Audio:** Start playing a silent audio clip on a user action and then programmatically control the playback of your actual audio clip.
- **Web Audio API:** Use the Web Audio API to gain more control over the audio playback and handle user interactions more effectively.

Here's a simplified example of how you might modify your code to start the audio on a user action:
```dart
AudioPlayer player = AudioPlayer();
bool isPlaying = false;

void playSound() {
  if (!isPlaying) {
    player.play(AssetSource('audio/68221786.mp3'));
    isPlaying = true;
  }
}

// In your button's onPressed callback:
onPressed: () {
  playSound();
  timer = Timer.periodic(const Duration(milliseconds: 2000), (timer) async {
    playSound();
  });
},
```
This example assumes that you have a button that, when pressed, starts the audio and sets up a periodic timer to play the sound every 2 seconds. The isPlaying flag is used to ensure that the audio starts playing in response to a user action.

Remember to test your solution on different devices and browsers to ensure compatibility. If you need further assistance or have more questions, feel free to ask.

**Reply by toutPetitPoissonRouge** [reply]
Thank you very much for you input and ideas. Nevertheless in you proposed code you are starting a periodic timer just exactly as i do it in my orignial code, which does not work. So i don't see how your proposition should solve the issue... can you give some explanation in more deepness ?
